### DISCUSSION TITLE
Breakout metadata into separate router segment (metadata.ts)

### DISCUSSION THREAD

**Discussion by BarakChamo** [vercel/next.js#63799] [App Router]
Hi there,

I'd like to make suggest a new file type for App Router route segments to encapsulate metadata separately from `route` and `layout` files, allowing for other SSR segments to be cached.

### Issue with metadata

According to Next.js documentation, two scenarios that will force a route segment to be dynamically rendered are:
1. using `searchparams` in `page.tsx` components.
2. using `cookies`

Both of these are frequently used to customize metadata for situations such as:
1. referrals and tracking
2. open graph and link preview generation
3. SEO

In many of our projects, metadata is often the only (or most frequent) dynamic part in a layout or page to change on the server.
Since metadata has to change on the server and cannot be extracted to a client component, this is forcing our pages to be dynamically rendered.

Also, since metadata is exported as structured data from a route segment, rather than say, a component, it cannot be easily extracted into a suspense boundary and take advantage of `ppr` composition.

### Example 

For example, imagine implementing a referral system that not only renders a dynamic OG image, but also customizes the title and text, based on how links are previewed on [messaging apps like Telegram](https://support.wix.com/en/article/updating-your-pages-link-preview-in-telegram).

You would want to customize the `og:*` meta tags with something like:
```json
{
   "og:title": "Hey Joe, you've invited!",
   "og: description": "Steve sent you a special invite to join Acme Inc."
}
```

Even if this is the only part that is dynamic on the server, this would require parsing searchparams which would force the whole route to be dynamic.

### Proposal

It would be nice to have a new type of route segment, `metadata.ts` or `headers.ts` that would allow us to split the route layouts and pages from the metadata generation which is more often dynamic.

A similar thing is already possible with open-graph file conventions for dynamic open-graph images: `opengraph-image.tsx` and `twitter-image`.tsx

**Comment by itsjavi** [comment]
I was also trying to find `metadata.ts` in the [documentation](https://nextjs.org/docs/getting-started/project-structure), but I realized we don't have that yet.

It totally makes sense for me to have this new Routing file, with the default export being the metadata object or a metadata generator function.

Another reason for that are routes that need to define a metadata template but don't need necessarily a new layout. I have many files that are just a wrapper layout because of this, e.g.

```tsx
// layout.tsx

// Useless wrapper. If omitted we get the error "Error: The default export is not a React Component"
export default function Layout({children}: Readonly<{children: React.ReactNode}>) {
  return children
}

// Template metadata for all child routes:
export const metadata: Metadata = {
  title: {
    template: `%s - ${metaTitle} - ${appConfig.siteName}`,
    default: metaTitle,
  },
  description: metaDescription,
  robots: 'index, follow',
  alternates: {
    canonical: fullUrl,
  },
  twitter: {
    card: 'summary_large_image',
    site: '@supereffectiv',
    title: metaTitle,
    description: metaDescription,
    images: [],
  },
  openGraph: {
    title: metaTitle,
    description: metaDescription,
    type: 'article',
    tags: ['pokemon', 'pokedex', 'dex'],
    url: fullUrl,
    images: [],
  },
}
```

**Comment by BarakChamo** [comment]
We're having to do the same since moving away from the `pages` router and `head.ts` file.

Even the root page is wrapped inside a `(base)` route group to nest another layout to address changes in metadata based on dynamic data. This is not only an inconvenience compared to the `pages` router, but **forces us into a next.js anti-pattern**.

### Here is an example of this anti-pattern:

Suppose, as in my original example:
1. I wanted to rely on `searchParams` to customize the content of `openGraph` values of a page's metadata
2. This is a content heavy page that would benefit from static rendering since the only thing changing is the metadata

Since `searchParams` are not exposed to `layout.tsx`, wrapping the page in a thin `layout` to handle the dynamic portion is not possible. 

Instead, I'm forced to move the content of the page into a nested layout inside a transparent `(route group)`, and have the page only set the metadata and render nothing else.

```ascii
app/
├─ layout.tsx - static
├─ (base)/
│  ├─ page.tsx - dynamic
│  ├─ layout.tsx static (actual page content)
```

This is very confusing and potentially has other performance or SEO implications I'm unaware of.

Instead, it would be much cleaner and more "correct" to have a file tree like this:
```ascii
app/
├─ layout.tsx
├─ page.tsx
├─ metadata.ts
```
